src/node_contextify.cc

```diff
@@ -432,7 +432,10 @@ class ContextifyContext {
     // false for vmResult.x = 5 where vmResult = vm.runInContext();
     bool is_contextual_store = ctx->global_proxy() != args.This();
 
-    if (!is_declared && args.ShouldThrowOnError() && is_contextual_store)
+    bool is_function = value->IsFunction();
+
+    if (!is_declared && args.ShouldThrowOnError() && is_contextual_store
+    && !is_function)
```

Comment:
@bnoordhuis, thanks! I realized I should have clarified that this applies to the strict mode (edited the commits). Outside of the strict mode ```args.ShouldThrowOnError()=false ```, so this does not apply.
When in the strict mode, the only change in behavior happens for ```function f() {}```, where now we pass the condition: it is not declared, nor defined and "we are strict", so without checking that it might be a function declaration, it returns and never calls ```ctx->sandbox()->Set(property, value);```, relying on ```CopyProperties()``` to set ```f``` on the sandbox.  
For ```var f = function()```, ``` is_declared = true``` and 
for ```this.f = function() {}```  ```is_contextual_store=false```, so ```is_function``` does not make a difference.